fw/http.c

```diff
@@ -5347,6 +5388,39 @@ tfw_h2_resp_adjust_fwd(TfwHttpResp *resp)
 	if (unlikely(r))
 		goto clean;
 
+	if (resp->trailers_len > 0) {
+		// MUST _NOT_ use resp pool to encode trailer headers
+		// because it assumes no headers after body in many code.
+		// Instead, encode trailers header in new skbs
+		// and append them to the resp->msg.skb_head.
+		struct sk_buff *skb_head, *skb, *nskb_head, *nskb;
+		unsigned long acc;
+
+		skb_head = resp->msg.skb_head;
+		skb = resp->mit.iter.skb;
+		resp->msg.skb_head = resp->mit.iter.skb = NULL;
+		acc = resp->mit.acc_len;
+
+		r = tfw_h2_hpack_encode_trailer_headers(resp, h_mods, stream_id);
+
+		resp->req->stream->xmit.t_len = resp->mit.acc_len - acc;
+		nskb_head = resp->msg.skb_head;
+		resp->msg.skb_head = resp->mit.iter.skb = skb;
+
+		if (unlikely(r))
+			goto clean;
+
+		nskb = nskb_head;
```

Comment:
We should set appropriate skb flags in __tfw_hpack_encode, not implement additional loop here.

Reply:
No, here we no only need to handle the flags but also the skb list (for trailer headers). `__tfw_hpack_encode` has no information here. Since the new skb list is wrapped on the original code path, so we need to iterate all skbs and attach them to the message's skb head.